keanu-project/src/main/java/io/improbable/keanu/plating/Plate.java

```diff
@@ -1,23 +1,58 @@
 package io.improbable.keanu.plating;
 
-import io.improbable.keanu.vertices.Vertex;
-
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.stream.Collectors;
+
+import io.improbable.keanu.tensor.Tensor;
+import io.improbable.keanu.vertices.ProxyVertex;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.VertexDictionary;
+import io.improbable.keanu.vertices.VertexLabel;
+
+public class Plate implements VertexDictionary {
 
-public class Plate {
-    private Map<String, Vertex<?>> contents;
+    private Map<VertexLabel, Vertex<?>> contents;
 
     public Plate() {
         this.contents = new HashMap<>();
     }
 
-    public <T extends Vertex<?>> T add(String name, T v) {
-        contents.put(name, v);
+    public <T extends Vertex<?>> T add(T v) throws PlateException {
+        if (v.getLabel() == null) {
+            throw new PlateException("Vertex " + v + " has no label");
+        }
+        VertexLabel label = scoped(v.getLabel());
+        if (contents.containsKey(label)) {
+            throw new IllegalArgumentException("Key " + label + " already exists");
+        }
+        contents.put(label, v);
+        v.setLabel(label);
         return v;
     }
 
-    public <T> Vertex<T> get(String name) {
-        return (Vertex<T>) contents.get(name);
+    private String getUniqueName() {
+        return "Plate_" + this.hashCode();
+    }
+
+    private VertexLabel scoped(VertexLabel label) {
+        return label.withExtraNamespace(getUniqueName());
+    }
+
+    @Override
+    public <V extends Vertex<? extends Tensor<?>>> V get(VertexLabel label) {
+        Vertex<?> vertex = contents.get(label);
```

Comment:
Yes that's neat, although I think I will do it the other way round, so that it's less likely to fail and create a new entry in the map:

        Vertex<?> vertex = contents.computeIfAbsent(scoped(label), (l) -> contents.get(label));

Reply:
Good idea - although actually a `getOrDefault` is better since it doesn't mutate `contents`.

        Vertex<?> vertex = contents.getOrDefault(label, contents.get(scoped(label)));